### PR TITLE
[FLINK-26783] Do not trigger global failover if failed during commiting side-effects during stop-with-savepoint

### PR DESCRIPTION
**Pull request by dawidwys**
## What is the purpose of the change

If a job fails during commitin side-effects of the stop-with-savepoint and we restart to the latest checkpoint instead of savepoint, we might end up producing duplicates. On the other hand if we restart to the savepoint we end up in a situation where a running Flink job depends on the existence of the savepoint.

In this commit we do not trigger a global failover in case the savepoint completed successfully, but the job failed during committing side effects. In that case we will finish the completable future with an exception that explains that the savepoint is consistent, but it might have uncommitted side effects and we ask users to manually restart a job from that savepoint if they want to commit side effects.

## Verifying this change

Added test in:
* SavepointITCase

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (**yes** / no / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by gaoyunhaii** [comment]
@flinkbot run azure

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java [hunk 1]
```diff
@@ -1243,7 +1243,7 @@ private void completePendingCheckpoint(PendingCheckpoint pendingCheckpoint)
             // the pending checkpoint must be discarded after the finalization
             Preconditions.checkState(pendingCheckpoint.isDisposed() && completedCheckpoint != null);
 
-            if (!props.isSavepoint()) {
+            if (!props.isSavepoint() || props.isSynchronous()) {
                 lastSubsumed =
                         addCompletedCheckpointToStoreAndSubsumeOldest(
```
Comment (pnowojski) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java:1248]: Should we subsume anything in this case? Also isn't this braking savepoint ownership?

Comment (dawidwys) [reply flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java:1248]: 1. For the subsumption, we do have a logic that will keep at least one other checkpoint (not-savepoint), so I think that's fine. (see: `CheckpointSubsumeHelper#subsume`).
2. As for the ownership, there are two aspects here. First, we can say savepoint (stop-with-savepoint) is under Flink's control until the job has finished. Secondly, this reverts to the behaviour of <=1.14, the better handling we discussed should be implemented in https://issues.apache.org/jira/browse/FLINK-26683

**Comment by dawidwys** [comment]
After an offline discussion we said that indeed this approach poses a problem for the savepoint ownership, as after a restart the savepoint will remain in the `CompletedCheckpointStore` and Flink will depend on its existence.

Therefore we propose a different approach to solve the issue that if we fallback to a checkpoint we might end up with duplicated records. We suggest to already not trigger a global failover in case the savepoint completed successfully, but the job failed during committing side effects. In that case we will finish the completable future with an exception that explains that the savepoint is consistent, but it might have uncommitted side effects and ask users to manually restart a job from that savepoint if they want to commit side effects.

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/stopwithsavepoint/StopWithSavepointTerminationHandlerImpl.java [hunk 2]
```diff
@@ -167,16 +166,13 @@ private void handleAnyExecutionNotFinished(Set<ExecutionState> notFinishedExecut
...
         log.warn(
-                "A savepoint was created at {} but the corresponding job {} didn't terminate successfully.",
-                savepointPath,
-                jobId,
+                "Inconsistent execution state after stopping with savepoint. At least one"
+                        + " execution is still in one of the following states: {}.",
+                StringUtils.join(unfinishedExecutionStates, ", "),
                 inconsistentFinalStateException);
 
         scheduler.handleGlobalFailure(inconsistentFinalStateException);
```
Comment (pnowojski) [inline flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/stopwithsavepoint/StopWithSavepointTerminationHandlerImpl.java:178]: Shouldn't we change something around the failover behaviour?

**Review by pnowojski — APPROVED** [review]
Thanks for the explanation. I think LGTM

**Pull request closed without merging**